File: ios/Core/Sources/PrivycsCore/Pool/PoolHostnameLabels.swift
import Foundation

/// Hostname-pattern labels: city + country + flag for VPN provider
/// hostnames like "at-vie-wg-001". Field-for-field port of Android's
/// `PoolHostnameLabels` (same curated tables). Naming convention is
/// "<cc>-<city3>-<protocol>-<n>" where <city3> is an IATA airport code.
public enum PoolHostnameLabels {

    /// City full name from "<cc>-<city3>-..." (else "").
    public static func cityFromHostname(_ name: String?) -> String {
        guard let name, !name.isEmpty else { return "" }
        let parts = name.split(separator: "-", omittingEmptySubsequences: false)
        guard parts.count >= 2 else { return "" }
        return cityCodes[parts[1].lowercased()] ?? ""
    }

    /// Country full name from ISO-3166-1 alpha-2 (else ""), localized to the
    /// given locale. CLDR ships every country name in all our languages, so we
    /// prefer the OS localization (pass the in-app-language locale to honor the
    /// user's language pick) and fall back to the curated English table for any
    /// code CLDR can't resolve.
    public static func countryNameFromCode(_ cc: String?, locale: Locale = .current) -> String {
        guard let cc, !cc.isEmpty else { return "" }
        let code = cc.uppercased()
        if let localized = locale.localizedString(forRegionCode: code), !localized.isEmpty {
            return localized
        }
        return countryNames[code] ?? ""
    }

    /// Country flag emoji via Regional Indicator Symbols (else "").
    public static func flagEmoji(_ cc: String?) -> String {
        guard let cc, cc.count == 2 else { return "" }
        let upper = cc.uppercased()
        guard upper.unicodeScalars.allSatisfy({ $0.value >= 65 && $0.value <= 90 }) else { return "" }
        var out = ""
        for u in upper.unicodeScalars {
            if let scalar = UnicodeScalar(0x1F1E6 + (Int(u.value) - 65)) {
                out.unicodeScalars.append(scalar)
            }
        }
        return out
    }

    private static let cityCodes: [String: String] = [
        "vie": "Vienna", "fra": "Frankfurt", "ber": "Berlin", "muc": "Munich",
        "dus": "Düsseldorf", "ham": "Hamburg", "zrh": "Zurich", "gva": "Geneva",
        "par": "Paris", "mrs": "Marseille", "lon": "London", "mnc": "Manchester",
        "glw": "Glasgow", "mad": "Madrid", "bcn": "Barcelona", "mil": "Milan",
        "rom": "Rome", "ams": "Amsterdam", "bru": "Brussels",
        "sto": "Stockholm", "got": "Gothenburg", "mma": "Malmö", "osl": "Oslo",
        "cph": "Copenhagen", "hel": "Helsinki",
        "prg": "Prague", "war": "Warsaw", "buh": "Bucharest", "sof": "Sofia",
        "bud": "Budapest", "ath": "Athens", "lis": "Lisbon", "dub": "Dublin",
        "tll": "Tallinn", "rix": "Riga", "vno": "Vilnius", "beg": "Belgrade",
        "zag": "Zagreb", "lju": "Ljubljana", "bts": "Bratislava", "kiv": "Kyiv",
        "nyc": "New York", "chi": "Chicago", "lax": "Los Angeles", "sea": "Seattle",
        "sjc": "San Jose", "mia": "Miami", "dal": "Dallas", "den": "Denver",
        "atl": "Atlanta", "phx": "Phoenix", "bos": "Boston", "iad": "Washington",
        "slc": "Salt Lake City",
        "yyz": "Toronto", "yvr": "Vancouver", "ymq": "Montreal", "mex": "Mexico City",
        "sao": "São Paulo", "gru": "São Paulo", "eze": "Buenos Aires", "scl": "Santiago",
        "bog": "Bogotá", "lim": "Lima",
        "tok": "Tokyo", "nrt": "Tokyo", "osa": "Osaka", "sel": "Seoul", "icn": "Seoul",
        "hkg": "Hong Kong", "tpe": "Taipei",
        "sin": "Singapore", "kul": "Kuala Lumpur", "bkk": "Bangkok", "jkt": "Jakarta",
        "mnl": "Manila", "hnd": "Hanoi", "sgn": "Ho Chi Minh City",
        "bom": "Mumbai", "del": "Delhi", "blr": "Bangalore",
        "syd": "Sydney", "mel": "Melbourne", "per": "Perth", "akl": "Auckland",
        "jnb": "Johannesburg", "cpt": "Cape Town", "lag": "Lagos", "nai": "Nairobi",
        "cai": "Cairo",
        "dxb": "Dubai", "tlv": "Tel Aviv", "ist": "Istanbul",
    ]

    private static let countryNames: [String: String] = [
        "AT": "Austria", "BE": "Belgium", "BG": "Bulgaria", "HR": "Croatia",
        "CZ": "Czechia", "DK": "Denmark", "EE": "Estonia", "FI": "Finland",
        "FR": "France", "DE": "Germany", "GR": "Greece", "HU": "Hungary",
        "IE": "Ireland", "IT": "Italy", "LV": "Latvia", "LT": "Lithuania",
        "LU": "Luxembourg", "MT": "Malta", "NL": "Netherlands", "PL": "Poland",
        "PT": "Portugal", "RO": "Romania", "SK": "Slovakia", "SI": "Slovenia",
        "ES": "Spain", "SE": "Sweden", "GB": "United Kingdom", "UK": "United Kingdom",
        "CH": "Switzerland", "NO": "Norway", "IS": "Iceland", "AL": "Albania",
        "BA": "Bosnia", "MK": "North Macedonia", "ME": "Montenegro", "RS": "Serbia",
        "MD": "Moldova", "UA": "Ukraine", "BY": "Belarus", "RU": "Russia",
        "TR": "Turkey", "XK": "Kosovo", "LI": "Liechtenstein", "MC": "Monaco",
        "SM": "San Marino", "AD": "Andorra",
        "US": "United States", "CA": "Canada", "MX": "Mexico",
        "BR": "Brazil", "AR": "Argentina", "CL": "Chile", "CO": "Colombia",
        "PE": "Peru", "VE": "Venezuela", "EC": "Ecuador", "BO": "Bolivia",
        "PY": "Paraguay", "UY": "Uruguay",
        "JP": "Japan", "KR": "South Korea", "CN": "China", "TW": "Taiwan",
        "HK": "Hong Kong", "MO": "Macau", "TH": "Thailand", "VN": "Vietnam",
        "PH": "Philippines", "ID": "Indonesia", "MY": "Malaysia", "SG": "Singapore",
        "IN": "India", "PK": "Pakistan", "BD": "Bangladesh", "AU": "Australia",
        "NZ": "New Zealand",
        "ZA": "South Africa", "EG": "Egypt", "NG": "Nigeria", "KE": "Kenya",
        "MA": "Morocco", "DZ": "Algeria", "TN": "Tunisia",
        "AE": "UAE", "SA": "Saudi Arabia", "IL": "Israel", "QA": "Qatar",
        "KW": "Kuwait", "BH": "Bahrain", "OM": "Oman", "JO": "Jordan",
        "LB": "Lebanon", "IR": "Iran", "IQ": "Iraq",
        "DO": "Dominican Republic", "JM": "Jamaica", "BS": "Bahamas",
        "BB": "Barbados", "TT": "Trinidad", "CU": "Cuba", "PR": "Puerto Rico",
    ]
}

File: android/app/src/main/java/com/privycs/vpn/data/PoolHostnameLabels.kt
package com.privycs.vpn.data

/**
 * Hostname-pattern labels: city + country + flag for VPN provider
 * hostnames like "at-vie-wg-001" or "us-dal-wg-704".
 *
 * Port of desktop's pool_city_codes.go + ad-hoc country-name lookup.
 * The naming convention is "<cc>-<city3>-<protocol>-<n>" where
 * <city3> is typically an IATA airport code. Hand-curated list of
 * ~100 codes that actually appear across Mullvad, IVPN, Proton,
 * AzireVPN, PIA - shipping the full 17000+ IATA dump would bloat
 * the APK for zero practical benefit.
 *
 * Why this lives in `data/`: ConnectScreen + WidgetProvider both
 * read it. Putting it next to the Pool data avoids duplicating the
 * same lookup logic across multiple UI surfaces.
 */
object PoolHostnameLabels {

    /**
     * Parses "<cc>-<city3>-..." and returns the city full name, or
     * "" if the host doesn't match the pattern OR the city code is
     * not in the curated table. Mirrors desktop's
     * `cityFromHostnamePattern`.
     */
    fun cityFromHostname(name: String?): String {
        if (name.isNullOrEmpty()) return ""
        val parts = name.split("-")
        if (parts.size < 2) return ""
        return CITY_CODES[parts[1].lowercase()].orEmpty()
    }

    /**
     * Country full name from ISO-3166-1 alpha-2 code, localized to the
     * given locale (default = the active app/device locale), or "" if the
     * code is unknown. The JDK ships every country name in all our languages
     * via CLDR, so we prefer [java.util.Locale.getDisplayCountry] and fall
     * back to the curated English table only for codes CLDR can't resolve.
     */
    fun countryNameFromCode(cc: String?, locale: java.util.Locale = java.util.Locale.getDefault()): String {
        if (cc.isNullOrEmpty()) return ""
        val code = cc.uppercase()
        val display = java.util.Locale("", code).getDisplayCountry(locale)
        // getDisplayCountry echoes the code back when CLDR doesn't know it.
        if (display.isNotEmpty() && !display.equals(code, ignoreCase = true)) return display
        return COUNTRY_NAMES[code].orEmpty()
    }

    /**
     * Country flag emoji from ISO-3166-1 alpha-2 code, or "" if
     * the input is malformed. Uses the Regional Indicator Symbol
     * trick: each ASCII letter A-Z is offset to its corresponding
     * REGIONAL INDICATOR SYMBOL LETTER (U+1F1E6 + (letter - 'A')).
     * Android (API 26+) renders the pair as a flag glyph.
     *
     * No SVG assets, no Twemoji bundle - the OS font handles it.
     * Saves ~600KB of flag SVGs across the APK.
     */
    fun flagEmojiFromCode(cc: String?): String {
        if (cc.isNullOrEmpty() || cc.length != 2) return ""
        val upper = cc.uppercase()
        if (!upper.all { it in 'A'..'Z' }) return ""
        val first = 0x1F1E6 + (upper[0].code - 'A'.code)
        val second = 0x1F1E6 + (upper[1].code - 'A'.code)
        val sb = StringBuilder()
        sb.appendCodePoint(first)
        sb.appendCodePoint(second)
        return sb.toString()
    }

    private val CITY_CODES = mapOf(
        // Europe - DACH
        "vie" to "Vienna", "fra" to "Frankfurt", "ber" to "Berlin",
        "muc" to "Munich", "dus" to "Düsseldorf", "ham" to "Hamburg",
        "zrh" to "Zurich", "gva" to "Geneva",
        // Europe - West
        "par" to "Paris", "mrs" to "Marseille", "lon" to "London",
        "mnc" to "Manchester", "glw" to "Glasgow", "mad" to "Madrid",
        "bcn" to "Barcelona", "mil" to "Milan", "rom" to "Rome",
        "ams" to "Amsterdam", "bru" to "Brussels",
        // Europe - Nordic
        "sto" to "Stockholm", "got" to "Gothenburg", "mma" to "Malmö",
        "osl" to "Oslo", "cph" to "Copenhagen", "hel" to "Helsinki",
        // Europe - East
        "prg" to "Prague", "war" to "Warsaw", "buh" to "Bucharest",
        "sof" to "Sofia", "bud" to "Budapest", "ath" to "Athens",
        "lis" to "Lisbon", "dub" to "Dublin", "tll" to "Tallinn",
        "rix" to "Riga", "vno" to "Vilnius", "beg" to "Belgrade",
        "zag" to "Zagreb", "lju" to "Ljubljana", "bts" to "Bratislava",
        "kiv" to "Kyiv",
        // North America - US
        "nyc" to "New York", "chi" to "Chicago", "lax" to "Los Angeles",
        "sea" to "Seattle", "sjc" to "San Jose", "mia" to "Miami",
        "dal" to "Dallas", "den" to "Denver", "atl" to "Atlanta",
        "phx" to "Phoenix", "bos" to "Boston", "iad" to "Washington",
        "slc" to "Salt Lake City",
        // North America - Canada / Mexico
        "yyz" to "Toronto", "yvr" to "Vancouver", "ymq" to "Montreal",
        "mex" to "Mexico City",
        // South America
        "sao" to "São Paulo", "gru" to "São Paulo", "eze" to "Buenos Aires",
        "scl" to "Santiago", "bog" to "Bogotá", "lim" to "Lima",
        // Asia - East
        "tok" to "Tokyo", "nrt" to "Tokyo", "osa" to "Osaka",
        "sel" to "Seoul", "icn" to "Seoul", "hkg" to "Hong Kong",
        "tpe" to "Taipei",
        // Asia - South-East
        "sin" to "Singapore", "kul" to "Kuala Lumpur", "bkk" to "Bangkok",
        "jkt" to "Jakarta", "mnl" to "Manila", "hnd" to "Hanoi",
        "sgn" to "Ho Chi Minh City",
        // Asia - South
        "bom" to "Mumbai", "del" to "Delhi", "blr" to "Bangalore",
        // Oceania
        "syd" to "Sydney", "mel" to "Melbourne", "per" to "Perth",
        "akl" to "Auckland",
        // Africa
        "jnb" to "Johannesburg", "cpt" to "Cape Town", "lag" to "Lagos",
        "nai" to "Nairobi", "cai" to "Cairo",
        // Middle East
        "dxb" to "Dubai", "tlv" to "Tel Aviv", "ist" to "Istanbul"
    )

    private val COUNTRY_NAMES = mapOf(
        // Europe
        "AT" to "Austria", "BE" to "Belgium", "BG" to "Bulgaria",
        "HR" to "Croatia", "CZ" to "Czechia", "DK" to "Denmark",
        "EE" to "Estonia", "FI" to "Finland", "FR" to "France",
        "DE" to "Germany", "GR" to "Greece", "HU" to "Hungary",
        "IE" to "Ireland", "IT" to "Italy", "LV" to "Latvia",
        "LT" to "Lithuania", "LU" to "Luxembourg", "MT" to "Malta",
        "NL" to "Netherlands", "PL" to "Poland", "PT" to "Portugal",
        "RO" to "Romania", "SK" to "Slovakia", "SI" to "Slovenia",
        "ES" to "Spain", "SE" to "Sweden", "GB" to "United Kingdom",
        "UK" to "United Kingdom", "CH" to "Switzerland", "NO" to "Norway",
        "IS" to "Iceland", "AL" to "Albania", "BA" to "Bosnia",
        "MK" to "North Macedonia", "ME" to "Montenegro", "RS" to "Serbia",
        "MD" to "Moldova", "UA" to "Ukraine", "BY" to "Belarus",
        "RU" to "Russia", "TR" to "Turkey", "XK" to "Kosovo",
        "LI" to "Liechtenstein", "MC" to "Monaco", "SM" to "San Marino",
        "AD" to "Andorra",
        // North America
        "US" to "United States", "CA" to "Canada", "MX" to "Mexico",
        // South America
        "BR" to "Brazil", "AR" to "Argentina", "CL" to "Chile",
        "CO" to "Colombia", "PE" to "Peru", "VE" to "Venezuela",
        "EC" to "Ecuador", "BO" to "Bolivia", "PY" to "Paraguay",
        "UY" to "Uruguay",
        // Asia-Pacific
        "JP" to "Japan", "KR" to "South Korea", "CN" to "China",
        "TW" to "Taiwan", "HK" to "Hong Kong", "MO" to "Macau",
        "TH" to "Thailand", "VN" to "Vietnam", "PH" to "Philippines",
        "ID" to "Indonesia", "MY" to "Malaysia", "SG" to "Singapore",
        "IN" to "India", "PK" to "Pakistan", "BD" to "Bangladesh",
        "AU" to "Australia", "NZ" to "New Zealand",
        // Africa
        "ZA" to "South Africa", "EG" to "Egypt", "NG" to "Nigeria",
        "KE" to "Kenya", "MA" to "Morocco", "DZ" to "Algeria",
        "TN" to "Tunisia",
        // Middle East
        "AE" to "UAE", "SA" to "Saudi Arabia", "IL" to "Israel",
        "QA" to "Qatar", "KW" to "Kuwait", "BH" to "Bahrain",
        "OM" to "Oman", "JO" to "Jordan", "LB" to "Lebanon",
        "IR" to "Iran", "IQ" to "Iraq",
        // Caribbean
        "DO" to "Dominican Republic", "JM" to "Jamaica", "BS" to "Bahamas",
        "BB" to "Barbados", "TT" to "Trinidad", "CU" to "Cuba", "PR" to "Puerto Rico"
    )
}
